README
-*-Text-*-
____________
Introduction

Calife est un programme permettant à certains utilisateurs, dont les noms sont
présents dans une liste, de passer sous un quelconque ID utilisateur (par
exemple, mais pas uniquement, "root") en rentrant leur propre mot de passe.

Ce système permet de ne pas divulguer le mot de passe de root (par exemple
pour le mode "single-user" de certains Unix) tout en conférant les droits de
root à certains utilisateurs bien identifiés.

L'introduction obligatoire du mot de passe lors du changement d'ID permet de
ne pas compromettre l'accès root si l'utilisateur reste logé sur une machine
puis s'en éloigne.

La barrière de sécurité finale réside dans le mot de passe de ces
utilisateurs, qui, nous l'espérons, connaissent les règles élémentaires.

Des listes d'utilisateurs vers lesquels quelqu'un a le droit d'utiliser calife
permettent d'affiner la gestion des personnes autorisées.

____________
Distribution

Calife est distribué sous la GNU General Public Licence qui vous permet de
diffuser, utiliser et modifier ce programme tant que vous en diffusez les
sources et me communiquez toutes les modifications effectuées. Cela me permet
de l'améliorer et de le maintenir dans de bonnes conditions.

Vous trouverez la licence GNU dans le fichier COPYING fourni avec l'archive.

Pour ce qui est du "disclaimer" standard associé à ce type de programme et à
ce type de diffusion, le paragraphe suivant, tiré de la licence BSD,
s'applique :

  THIS SOFTWARE IS PROVIDED BY ME ``AS IS'' AND ANY EXPRESS OR IMPLIED
  WARRANTIES, INCLUDING, BUT NOT LIMITED TO, THE IMPLIED WARRANTIES OF
  MERCHANTABILITY AND FITNESS FOR A PARTICULAR PURPOSE ARE DISCLAIMED. IN NO
  EVENT SHALL I BE LIABLE FOR ANY DIRECT, INDIRECT, INCIDENTAL, SPECIAL,
  EXEMPLARY, OR CONSEQUENTIAL DAMAGES (INCLUDING, BUT NOT LIMITED TO,
  PROCUREMENT OF SUBSTITUTE GOODS OR SERVICES; LOSS OF USE, DATA, OR PROFITS;
  OR BUSINESS INTERRUPTION) HOWEVER CAUSED AND ON ANY THEORY OF LIABILITY,
  WHETHER IN CONTRACT, STRICT LIABILITY, OR TORT (INCLUDING NEGLIGENCE OR
  OTHERWISE) ARISING IN ANY WAY OUT OF THE USE OF THIS SOFTWARE, EVEN IF
  ADVISED OF THE POSSIBILITY OF SUCH DAMAGE.

Ceci dit, si vous l'utilisez et le trouvez utile, je ne refuserais pas une
Guinness :-)

Le site primaire pour Calife est <http://www.keltia.net/programs/calife> et
en FTP :

<ftp://mutt.frmug.org/pub/calife/>. Un miroir existe sur
<ftp://postfix.EU.org/pub/calife/>.

Calife utilise maintenant Redmine (<http://redmine.org/>) comme frontal de
gestion des tickets, flux RSS, release, forums, etc. Le site est accessible à
<http://dev.keltia.net/projects/calife>.

Calife est également disponible sous la forme d'un dépôt Mercurial
<http://selenic.com/mercurial/> ce qui facilitera l'usage et la mise à jour
par des contributeurs externes.

Répertoire de base des dépôts Mercurial :

 <http://archives.keltia.net/hg/>

____________
Installation

Calife utilise dorénavant le programme GNU "autoconf" pour la configuration.
GNU cc n'est pas obligatoire pour compiler Calife mais ne sont testés
régulièrement que les compilateurs suivants :

- GNU cc 3.x et 4.3
- Clang 1.0, 1.1 et la version de développement (trunk)
- SUN cc sur plateforme Sparc

Tapez la commande :

./configure

Le script va analyser votre système et créer les fichiers suivants :

- Makefile
- config.h
- calife.1
- calife.auth.5

Le binaire et  les pages de  manuels seront par défaut installés, lors du
"make install", sous "/usr/local".

Utiliser l'argument de configure "--prefix=/le/prefixe" pour indiquer le nom
du répertoire adéquat.

Quelques options spéciales sont disponibles  :

  --with-etcdir=PATH        Directory containing calife.auth (default /etc)
  --disable-shadow=yes|no   Disable getspnam(3) usage for some Linuxes
  --enable-debug=yes|no     Set the debugging mode (default is no)
  --enable-shellhack=yes|no (default is no)
  --enable-dynamic          Build the program dynamically (default static)
  --enable-global-rc        Add /etc/calife.out support.


--with-etcdir=CHEMIN
        Permet de  placer les fichiers  calife.auth et  calife.out ailleurs
        que dans /etc. 

--enable-debug
        Exige l'affichage des messages de "debugging".
        Inutile en usage normal. 

--enable-shellhack
        Si vous avez le message "Permission denied" après avoir tapé votre
        mot de passe, essayez de relancer "configure" avec cette option. 

--enable-dynamic
        Si vous désirez avoir le binaire lié dynamiquement et non pas
        statiquement. L'utilisation du binaire statique permet non seulement
        d'améliorer le partage des pages entre de multiples instances de
        calife mais aussi de ne pas être dépendant de la version de la
        bibliothèque standard.

        Parmi les systèmes devant utiliser --enable-dynamic, on compte
        Solaris 2.x, HP-UX, Linux et MacOS X/Darwin.

Cas de Linux :

Ce système utilise plusieurs systèmes de "shadow passwords". Il est possible
que "configure" ne parvienne pas à déterminer le type d'outil de "shadow
passwords" employé. Je n'ai pas de Linux, il m'est donc difficile de tester...
Toute suggestion et modification de "configure.in" appréciées.

L'option suivante permet à calife de ne pas utiliser getspnam(3) dans le cas
où la fonction existe mais les shadows passwords ne sont pas utilisés.

--disable-shadow
        Le défaut est d'utiliser getspnam(3) si elle est définie.

Si vous devez créer une nouvelle section dans conf.h ou une nouvelle entrée
dans Makefile.in ou configure.in , merci de me communiquer les modifications
pour incorporation dans les versions suivantes.

Le programme s'installe sous forme d'un binaire nommé "calife" pour des
raisons de facilité mais le nom historique est "...".

Il installe aussi un fichier de configuration d'exemple, appelé
"calife.auth-dist" et par défaut placé dans /etc

Consulter les pages de manuel fournies ("calife" et "calife.auth") pour en
apprendre davantage.

Le programme essaie de lancer le fichier calife.out (par défaut dans
/usr/local/etc sur FreeBSD et /etc ailleurs) à la sortie de "calife". Ceci
permet par exemple d'envoyer un mail à chaque utilisation de calife. Cette
possibilité est optionnelle et dépend de l'option « --enable-global-rc ».

_____________
Compatibilité

Calife a été testé et développé sur les systèmes UNIX suivants (les versions
testées peuvent être anciennes et devraient être retestées mais je n'ai plus
accès à pas mal de ces machines...) :

- FreeBSD 1 - 6.x   (mach. de développement de l'auteur)
- MacOS X 10.2 - 10.6.
- Linux 0.99+ et 1.* (1.2.* et 1.3.*), 2.*
- Solaris 2.x (et la plupart des SVR4)
- HP-UX 8.*, 9.*, 10.* et 11.*

Bien que Calife n'ait pas été testé par l'auteur sur NetBSD et OpenBSD, il est
fort probable du fait de la base commune de code, il n'y ait que peu de souci
à utiliser Calife sur ces systèmes, à part sans doute quelques différences
dans le sous-système PAM.

NOTE: la version 3.0 sera une version orienté PAM de manière à simplifier le
code, sa maintenance et la compatibilité avec les UNIX le supportant, le code
non PAM restant a priori utilisable.

Systèmes pour lesquels la compatibilité n'est pas/plus testée directement par
l'auteur, ceux-ci étant donc non supportés.

- SunOS 4.1.*
- SunOS 3.2
- NeXTStep 3
- SCO 3.2
- HP-UX 8.* et 9.*
- SVR4 NeWSOS 5.02
- Encore Gould NP1
- AIX 3.x & 4.x
- IRIX 5.x (probablement 6.x aussi)
- GNU Hurd (nouveau)

Pour calife 3.0/PAM, voir les exemple de fichier de configuration de PAM
dans le répertoire pam/ de l'archive. Placer ce fichier dans /etc/pam.d (ou
l'endroit demandé par votre implémentation de PAM) sous le nom « calife ».

Il existe à l'heure actuelle deux fichiers :
- macosx    pour les systèmes MacOS 10.3/Panther à 10.5/Leopard ;
- macosx10  pour Mac OS X Snow Leopard 10.6 ;
- freebsd5  pour les systèmes FreeBSD 5.x et supérieurs
- solaris   pour Solaris 8, 9 et 10 (à insérer dans pam.conf)

Pour utiliser LDAP/NIS et autres systèmes d'authentification, il faudra 
modifier le fichier PAM correspondant.

La fonctionnalité syslog(3) est maintenant obligatoire pour compiler calife.
Ca simplifie le code manière significative.

Des packages FreeBSD et Debian existent dorénavant grâce à Thomas Quinot.
Alain Thivillon a généré des packages Red-hat (RPM) pour 2.8.4b.

_______________
Pages de manuel

Les pages de manuel sont très succinctes et utilisent le système de macros mis
au point à Berkeley appelé "mandoc" donc si vous n'avez pas ces macros, le
résultat n'est pas très lisible. Je conseille vivement d'utiliser ce système
qui est plus complet que le ``man'' habituel.

Si vous disposez de GNU groff, alors la commande pour les pages de manuel est
"nroff -mandoc" ou "gnroff -mandoc".

__________
Historique

Ce programme est né en 1991 d'une idée commune à François Berjon et moi, qui
utilisions un système très simple, appelé ``...'', capable de lancer un shell
root après vérification de l'uid du requérant, sans demander de mot de passe.

Il a évolué lors de mon travail à Antenne 2 et surtout lors de mon stage à
l'Université de Marne-la-Vallée, où il a gagné la majorité de ses possibilités
actuelles.

Il est utilisé quotidiennement par plusieurs sites et donne des résultats
satisfaisants.

Tout commentaire, critique [constructive] et patches bienvenus à l'une des
adresses suivantes :

roberto@keltia.net       (adresse privée maison)
Site: http://bitbucket.org/keltia/calife/

_____________
Remerciements

René Cougnenc           
François Berjon         <fcb@gwynedd.frmug.org>
Laurent Chemla          <laurent@brainstorm.fr>
Pierre Beyssac          <pb@fasterix.frmug.org>
Nathanaël Makarévitch   <nat@nataa.frmug.org>
Thomas Quinot           <thomas@cuivre.fr.eu.org>
Christian Perrier       <bubulle@kheops.frmug.org>
Philippe Regnault       <regnauld@starbsd.org>
Thierry Besancon        <Thierry.Besancon@csi.uvsq.fr>
Laurent Wacrenier       <lwa@teaser.fr>
Erwan David             <erwan@rail.eu.org>
Bertrand Petit          <elrond@phoe.frmug.org>
Denis Ducamp            <Denis.Ducamp@hsc.fr>

et à tous ceux qui ont participé pour la mise au point et les tests.

# $Id: README,v f71e70cf2d35 2013/05/11 12:40:30 roberto $


ChangeLog
2017-08-09  Ollivier Robert  <roberto@keltia.net>

	* README-fr.md, README.md:
	Sync with current state.
	[d5d1403715ea] [tip]

	* ChangeLog:
	Sync ChangeLog.
	[55c2817c0c2c]

	* calife.c:
	Wrap fd declaration with #ifdef WANT_GLOBAL_RC.
	[2b171253cc39]

	* calife.c:
	Remove trailing spaces.
	[9190b44575db]

	* calife.c:
	Separate the two cases to avoid unused variables..
	[ec92424830ea]

	* calife.c:
	Refactor: create whoami() to simplify main().
	[907189024d8f]

	* conf.h:
	Hook up compiler.h and use it.
	[f9fd34e270bf]

	* compiler.h:
	Include compiler.h by Bertrand Petit. Add keyword for declarations.
	[dbb4a7f658f1]

	* auth.c:
	Add assertions.
	[e5ce664a97ca]

	* auth.c:
	Test 1st char instead.
	[d6119254e1af]

	* calife.c:
	This is not necessary, exec() is coming.
	[225a2044ea10]

	* calife.c:
	Be more explicite.
	[e21871d017e0]

2017-08-08  Ollivier Robert  <roberto@keltia.net>

	* calife.c:
	Displays the strerror(3) output when fork(2) fails.

	Proposed by: Elrond <elrond@phoe.frmug.org>
	[7b2576d3600f]

2014-11-10  Ollivier Robert  <roberto@keltia.net>

	* .hgtags:
	Added tag 3.0.5 for changeset 12b49af8a87b
	[c9e854d492e5]

	* ChangeLog:
	Update.
	[12b49af8a87b] [3.0.5]

	* .hgtags:
	Added tag 3.0.5 for changeset 2200d81bf270
	[2bcc3cffc133]

	* .hgtags:
	Added tag 3.0.5 for changeset ee46e1fd4642
	[2200d81bf270]

	* config.guess, config.sub:
	Update config.guess/sub to 2014/11/04 versions.
	[ee46e1fd4642]

	* README-fr.md, README.md, pam/freebsd9:
	Add a freebsd9 PAM configuration file for 9.x and up. Fix READMEs.

	Submitted by: Dag-Erling Smørgrav <des@freebsd.org>
	[f91e6350b459]

	* ChangeLog:
	Merge.
	[3a061fcd5584]

	* ChangeLog:
	Re-sync ChangeLog.
	[b03e3cef8195]

2014-09-05  Ollivier Robert  <roberto@keltia.net>

	* .hgtags:
	Added tag 3.0.4 for changeset 25570e2f1dfc
	[0b9c5e950384]

	* calife.auth.5.in, db.c:
	Add support for %group in calife.auth.
	[25570e2f1dfc] [3.0.4]

	* calife.c:
	Fix messages, add ssid to debug.
	[57ac191d6c4c]

	* util.c:
	Clear out memory zones securely.
	[0849dfe2ac8e]

2014-05-23  Ollivier Robert  <roberto@keltia.net>

	* .hgtags:
	Added tag 3.0.3 for changeset 5d141ecaf6a2
	[3d15d0eb220f]

	* ChangeLog:
	Sync.
	[5d141ecaf6a2] [3.0.3]

	* .hgtags:
	Added tag 3.0.3 for changeset e47593d1cbda
	[2ca74c88ca3a]

	* configure:
	Regenerate for 3.0.3, should fix staged dir in FreeBSD ports.
	[e47593d1cbda]

	* Makefile.in:
	Remove unneeded messages. Do not rename old binary.
	[62657218adf3]

2014-03-20  Ollivier Robert  <roberto@keltia.freenix.fr>

	* conf.h:
	Merge.
	[b0fe97d87286]

2013-03-25  Ollivier Robert  <roberto@keltia.net>

	* .hgignore:
	Add some directories from NetBeans.
	[8ca4940b1cd0]

2012-10-26  Ollivier Robert  <roberto@keltia.net>

	* conf.h:
	Add a prototype for auth_pam() (silence warning from clang).
	[3b78f912d08d]

2014-02-23  Ollivier Robert  <roberto@keltia.net>

	* .hgtags:
	Added tag 3.0.2 for changeset 162282762a82
	[62a59249f305]

	* auth.c:
	Fix botched test.
	[162282762a82] [3.0.2]

	* .hgtags:
	Added tag 3.0.2 for changeset 8454656b738e
	[640962346f12]

	* Makefile.in:
	-Wall is default.
	[8454656b738e]

	* auth.c:
	Remove one variable, merge two portions of code.
	[fa7c6a7fb291]

	* auth.c:
	Remove unused variables (2nd pass). Add cheking on calife argument.
	[43b3e8e3085a]

	* conf.h:
	Add prototype for auth_pam().
	[c7bd68780eea]

	* auth.c:
	Remove unused variables (1st round).
	[b0e19aa3c07e]

2013-05-11  Ollivier Robert  <roberto@keltia.net>

	* README, README-fr.md, README.en, README.md:
	Move everything into Markdown.
	[ab55a2307363]

	* README, README.en:
	Fix addresses/URL.
	[f71e70cf2d35]

2012-08-22  Ollivier Robert  <roberto@keltia.net>

	* ChangeLog:
	Sync.
	[0855d1578d53]

	* Automated merge with
	ssh://keltia.freenix.org//www/archives.keltia.net/htdocs/hg/calife-
	pam
	[2ad8a83a9e6f]

	* db.c:
	Remove warning about (( )) from clang.
	[29ff9dbb76a9]

2012-01-03  Ollivier Robert  <roberto@keltia.net>

	* calife.auth-dist:
	We do not want some people to be feeling left out :).
	[c1cf7f18fe92]

2009-12-21  Ollivier Robert  <roberto@keltia.freenix.fr>

	* .hgtags:
	Added tag 3.0.1 for changeset f28c4191b823
	[7df82707b063]

	* ChangeLog:
	Updated.
	[f28c4191b823] [3.0.1]

	* README, README.en, pam/macosx10:
	OS X 10.6 has different PAM modules.
	[3c053a655692]

2009-12-11  Ollivier Robert  <roberto@keltia.freenix.fr>

	* configure:
	Regenerate.
	[933e238e8aa6]

2009-12-03  Ollivier Robert  <roberto@keltia.freenix.fr>

	* configure.in:
	Add a missing ;;.
	[d58abe72e2af]

	* configure:
	Regenerate.
	[5dc09f855c8c]

2009-12-02  Ollivier Robert  <roberto@keltia.freenix.fr>

	* config.guess, config.sub:
	Update these to 2009-11-19 versions.
	[bc7001ab775f]

	* config.guess, config.sub:
	Update these to 2008 versions.
	[5c531766b5c9]

	* calife.c:
	Fix copyright.
	[c67ec95b1aef]

2009-12-01  Ollivier Robert  <roberto@keltia.freenix.fr>

	* calife.c:
	Remove space after \.
	[0b340bd3c882]

	* calife.c:
	Reorder and reformat.
	[112c396b5525]

	* calife.c:
	Reorder.
	[5c7f2d9fb968]

	* calife.c:
	Remove unnecessary #ifdef.
	[a9a640c3c893]

	* calife.c:
	Untabify.
	[b6725dd2e87f]

	* calife.c:
	Remove potential NULL reference. Found by clang static analyzer.
	[51b770f23683]

	* calife.c:
	fp is no more here.
	[dceb74c03dbc]

	* ChangeLog:
	Updated.
	[b5e2803850c5]

	* db.c:
	Clean up open_databases(), remove access(2).
	[5a1f3f02b317]

	* conf.h:
	Tabs to spaces.
	[2db16132e2bd]

	* conf.h:
	Complete move of fp into db.c.
	[73677a6dc731]

	* calife.c, conf.h, db.c:
	Promote fp from global variable to static variable in db.c.
	[cfe779015a7e]

	* auth.c, calife.c, conf.h, db.c, util.c:
	Update copyright.
	[0dcc27d2f900]

	* ChangeLog:
	Updated. Re-encoded in UTF-8.
	[470097df63d5]

	* README, README.en:
	Updated & synced together. Now UTF-8 as well.
	[2398eb5baa35]

	* ChangeLog-2.x:
	Re-encode as UTF-8.
	[5e5862be12dd]

	* calife.c, util.c:
	Re-encode as UTF-8.
	[9c712aa88821]

	* configure.in:

db.c
/** Fichier db.c
 **
 ** Routines de gestion des mots de passe et des fichiers
 ** de configuration pour calife.
 **
 ** Copyright (c) 1991-2010 by Ollivier ROBERT
 ** A distribuer selon les regles de la GNU GPL General Public Licence
 ** Voir le fichier COPYING fourni.
 **
 **/

#ifndef lint
static const char * rcsid = "@(#) $Id: db.c,v 25570e2f1dfc 2014/09/05 21:08:02 roberto $";
#endif

#include "config.h"     /* GNU configure */

                        /* fichier de configuration */
#include "conf.h"

static FILE * fp = NULL;    /* Moved from calife.c */

/** On cherche et ouvre la database locale.
 **
 ** Dans tous les cas, le fichier log est ouvert (il est local). Si syslogd(8)
 ** est actif on ouvre le "fichier" pour syslog(3). Le fichier de log devient
 ** maintenant dependant de NO_SYSLOG: s'il n'y a pas de syslog d'utilisable
 ** alors on ouvre le fichier de log.
 **
 ** Parametres :    neant
 **
 ** Retourne :      int     s'il y a un probleme renvoie 1 sinon 0.
 **/

int    
open_databases (void)
{
    int err;

    MESSAGE ("Opening databases...\n");
    /*
     * become root again
     */
    GET_ROOT;

    /*
     * open protected database
     */
    fp = fopen (AUTH_CONFIG_FILE, "r");
    if (fp == NULL)
    {
        syslog (LOG_AUTH | LOG_ERR, "No database in `%s', launching su...\n", AUTH_CONFIG_FILE);
        MESSAGE_1 ("No database in `%s'...\n", AUTH_CONFIG_FILE);                              
        return 1;
    }
    else 
    {
        /*
         * XXX Is this still necessary?
         */    
        err = fchmod (fileno (fp), 0400);
        if (err)
            die(6, "Unable to change '%s' permissions.", AUTH_CONFIG_FILE);
            /*NOTREACHED*/
    }
    /*
     * stay non root for a time
     */
    RELEASE_ROOT;
    /*
     * Success
     */
    return 0;
}

/** Verifie si l'utilisateur de nom name a le droit d'utiliser
 ** calife.
 **
 ** Line format is :
 **
 ** login[:][shell][:user1,user2,user3,...]
 ** @group[:][shell][:user1,user2,...]
 **
 ** Parametres :    name        char *  nom de l'utilisateur
 **                 user_to_be  char *  futur nom d'utilisateur
 **
 ** Retourne :      int     1 si l'utilisateur est autorise 0 sinon.
 **/

int 
verify_auth_info (char * name, char * user_to_be)
{
    int     allowed = 0;
    char    ** user_list = NULL;
    size_t  l_line = 0;
    int     nb_users = 0;
    char    * line, * ptr;
    char    * line_name, * line_shell;
    char    * group_name = NULL, * group = NULL, **p;
    struct group *gr_group;
    
    /*
     * let's search if user allowed or not
     */
#ifdef HAVE_WORKING_SYSCONF
    l_line = (size_t) sysconf (_SC_LINE_MAX);   /* returns long usually */
#else
    l_line = MAX_STRING;
#endif /* HAVE_WORKING_SYSCONF */    

    line = (char *) xalloc (l_line);
    while (!feof (fp))
    {
        int lastc;

        if (fgets (line, l_line, fp) == NULL)
            break;
        /*
         * remove trailing '\n' if present
         */
        lastc = strlen (line) - 1;
        if (line [lastc] == '\n')
            line [lastc] = '\0';
        /*
         * empty line or comment  or null login -- not allowed but ignored
         */
        if (!line || (*line) == '#' || (*line) == '\0')
            continue;
        MESSAGE_1 ("Line read = |%s|\n", line);
        line_name = line;
        /*
         * Look for a @ or % as first character (for groups)
         */
        if (*line == '@' || *line == '%')
        {
            group = strdup(line);
            group_name = strtok(group, ":");
            group_name++;           /* skip '@' */
            /*
             *  Look into /etc/groups (or its equivalent with get)
             */
             gr_group = (struct group *) getgrnam(group_name);
             if (gr_group == NULL)
             {
                 die(1, "No such group %s", group_name);
                 allowed = 0;
                 goto end;
             }
             for ( p = &gr_group->gr_mem[0]; *p ; p++ )
             {
                 MESSAGE_3 ("matching %s to %s in %s\n", name, *p, group_name);  
                 if (strcmp(name, *p) == 0)
                 {
                     MESSAGE_2 ("User %s Allowed through group %s\n", name, group_name);
                     _group = strdup (group_name);
                     strcpy (_group, group_name);
                     allowed = 2;
                     break;
                 }
             }
        }
        /*
         * Look for first ':'
         */
        ptr = strchr (line, ':');
        if (ptr == NULL)
        {
            /*
             * No shell and no user list
             */
            custom_shell = 0;
            
            /*
             * Bypass if already allowed through group
             */
            if (group_name != NULL && allowed == 1)
                goto escape;
            if (strcmp (line_name, name))   /* not us */
                continue;
            goto escape;            
        }
        /*
         * I got a ':', put a '\0' instead.
         */
        *ptr++ = '\0';
        /*
         * Do we have anything behind ?
         */
        if (*ptr == '\0')
            custom_shell = 0;

        MESSAGE_2 ("Current line_name = |%s| / name = |%s|\n", line_name,name);

        /*
         * Bypass if already allowed through group
         */
        if (group_name != NULL && allowed != 0)
            goto end;
        if (strcmp (line_name, name))   /* not us */
            continue;

        line_shell = ptr;
        /*
         * Look for another ':', maybe just after the first one
         */
        ptr = strchr (line_shell, ':');
        if (ptr == line_shell)
            custom_shell = 0;            
        /*
         * If we got another one, put a '\0' instead
         */
        if (ptr)
            *ptr++ = '\0';
        /*
         * Analyze the shell candidate
         */
        if (*(line_shell) == '*')   /* LOCKED ACCOUNT */
        {
            allowed = 0;
            goto end;
        }
        /*
         * look for pathname
         */
        if (*(line_shell) == '/')
        {
            /*
             * Is this a real shell ?
             */
            if (access (line_shell, R_OK | X_OK))
            {
                MESSAGE_1 (" access of |%s| failed, no custom_shell\n",
                           line_shell);
                custom_shell = 0;
            }
            else
            {
                custom_shell = 1;
                shell = (char *) xalloc (strlen (line_shell) + 1);
                strncpy (shell, line_shell, strlen (line_shell));
                shell [strlen (line_shell)] = '\0';
                MESSAGE_1 (" current line_shell = |%s|\n", line_shell);
            }
        }
        else
        {
            custom_shell = 0;
        }
        MESSAGE_1 (" custom_shell = |%d|\n", custom_shell);

        /*
         * Analyze the rest of the line
         */
        if (ptr)
        {
            MESSAGE_1 ("ptr = %s\n", ptr);
            /*
             * we've got a user list
             */
            if (*ptr)
            {
                char  * p, ** q;
                
                MESSAGE_1 (" current user_list = |%s|\n", ptr);
                /*
                 * the list should be user1,user2,...,user_n
                 */
                p = ptr;     /* count users # in the list */
                while (p && *p)
                    if (*p++ == ',')
                        nb_users++;
                if (*--p != ',')    /* last char is not a ',' */
                    nb_users++;

                MESSAGE_1 ("  users # = |%d|\n", nb_users);

                /*
                 * two passes to avoid fixed size allocations
                 */ 
                user_list = (char **) xalloc((nb_users + 1) * sizeof(char *));
                /*
                 * put pointers in user_list
                 */
                p = ptr;     
                q = user_list;
                /*
                 * 1st user is immediate
                 */
                (*q++) = p;
                while (p && *p)
                {
                    if (*p == ',')    /* found a ',' */
                    {
                        *p = '\0';
                        if (*++p == ',')
                            continue;
                        (*q++) = p++;     /* next char is user name */
                    }            
                    else
                        p++;        /* only regular letters */
                }            
            }
        }        
        /*
         * user_list is complete, now compare
         */
escape:
        if (user_list)
        {
            int i;
            
            for ( i = 0; i < nb_users; i++ )
            {
                MESSAGE_2(" comparing |%s| to |%s|\n", 
                          user_to_be, user_list [i]);
                if (!strcmp (user_list [i], user_to_be))
                {
                    allowed = 1;
                    goto end;   /* beuh but too many loops */
                }
            }            
        }
        else
        {
            allowed = 1;
            break;            
        }
    }
end:
    free (line);
    if (group)
        free(group);
    if (user_list)
        free (user_list);
    fclose (fp);
    if (geteuid() == 0)
        allowed = 1;
    MESSAGE_1 ("Exit from verify_auth_info with allowed=%d\n", allowed);
    return allowed;    
}

